fab/fabfile.py

```diff
@@ -107,66 +109,49 @@ def _require_target():
 
 
 class DeployMetadata(object):
-    def __init__(self, env):
-        self.env = env
+    def __init__(self):
         self.timestamp = datetime.datetime.utcnow().strftime('%Y-%m-%d_%H.%M')
         self._deploy_ref = None
         self._deploy_tag = None
-        self._git = sh.git.bake(_tty_out=False)
+        self._repo = GitHub().repository('dimagi', 'commcare-hq')
+        self._max_tags = 100
+        self._last_tag = None
 
     def tag_commit(self):
-        self._git.fetch("origin", "--tags")
-        pattern = "*{}*".format(self.env.environment)
-        self.last_tag = sh.tail(self._git.tag("-l", pattern), "-1").strip()
-
-        tag_name = "{}-{}-deploy".format(self.timestamp, self.env.environment)
-        msg = "{} deploy at {}".format(self.env.environment, self.timestamp)
-        self._git.tag(tag_name, "-m", msg, self.deploy_ref)
-        self._git.push("origin", tag_name)
+        pattern = ".*-{}-.*".format(env.environment)
+        for tag in self._repo.tags(self._max_tags):
+            if re.match(pattern, tag.name):
+                self._last_tag = tag.name
+                break
+
+        if not self._last_tag:
+            print yellow('Warning: No previous tag found in last {} tags for {}'.format(
+                self._max_tags,
+                env.environment
+            ))
+        tag_name = "{}-{}-deploy".format(self.timestamp, env.environment)
+        msg = "{} deploy at {}".format(env.environment, self.timestamp)
+        self._repo.create_tag(tag_name, msg, self.deploy_ref)
         self._deploy_tag = tag_name
 
-    def _check_deploy_commit(self, deploy_commit, code_branch):
-        try:
-            origin_commit = self._git("rev-parse", "origin/{}".format(code_branch)).strip()
-        except sh.ErrorReturnCode:
-            print yellow("Branch '{}' was not found on origin".format(code_branch))
-            if not console.confirm("Do you want to deploy it anyways?", default=False):
-                utils.abort('Aborting.')
-        else:
-            if deploy_commit != origin_commit:
-                print yellow("\n*********\n*WARNING*\n*********")
-                print ("Your local copy of '{}' differs from the version on "
-                       "'origin'.\n".format(code_branch))
-                print "Here's the latest local commit:",
-                print self._git.show(deploy_commit, "--quiet")
-                print "Here's the latest commit on 'origin':",
-                print self._git.show(origin_commit, "--quiet")
-                if not console.confirm("Are you sure you want to deploy the "
-                                       "local commit?", default=False):
-                    utils.abort('Check yourself before you wreck yourself.')
```

Comment:
I think this is kind of a moot point since we very rarely actually check, but I just don't like the idea of not knowing what you're deploying.

The old way was to deploy your local `master` (or other branch if specified), not your local `HEAD`, so it shouldn't matter what branch you're on, as long as master is up-to-date.  The idea there is that you pull latest masters, update the submodules, check, then deploy that specific commit.

Here it deploys github's latest master, which the deployer has less control over.  Maybe a compromise approach would be to prompt the user with the `$ git show` of whatever that commit is, so there's a chance to verify that nothing unexpected has entered the mix last-minute.

Again, I think the overwhelming majority of the time it's irrelevant (so this is non-blocking), but I do like forcing the deployer to be cognizant of what's being deployed.